runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java

```diff
@@ -538,8 +540,8 @@ public IPath getStateLocation(Bundle bundle, boolean create) throws IllegalState
 	}
 
 	public long getStateTimeStamp() {
-		PlatformAdmin admin = getPlatformAdmin();
-		return admin == null ? -1 : admin.getState(false).getTimeStamp();
+		return Arrays.stream(getBundleContext().getBundles()).map(bundle -> bundle.adapt(Module.class))
+				.filter(Objects::nonNull).mapToLong(Module::getLastModified).max().orElse(-1);
```

Comment:
Alternatively the following could be used.
```suggestion
		Bundle bundle = FrameworkUtil.getBundle(InternalPlatform.class);
		org.eclipse.osgi.container.Module module = bundle.adapt(org.eclipse.osgi.container.Module.class);
		if (module == null) {
			return 0;
		}
		Generation generation = (Generation) module.getCurrentRevision().getRevisionInfo();
		org.eclipse.osgi.storage.Storage storage = generation.getBundleInfo().getStorage();
		return storage.getModuleDatabase().getRevisionsTimestamp();
```
The only problem with that code is that it uses internal Equinox methods and this project is configured to mark such usages with errors.

Another alternative would be to add to Equinox' `ModuleContainer` a method:
```
public long getRevisionsTimestamp() {
	return moduleDatabase.getRevisionsTimestamp();
}
```
And then just call
```suggestion
		Bundle bundle = FrameworkUtil.getBundle(InternalPlatform.class);
		org.eclipse.osgi.container.Module module = bundle.adapt(org.eclipse.osgi.container.Module.class);
		if (module == null) {
			return 0;
		}
		return module.getContainer().getRevisionsTimestamp();
```
All of that would have the advantage that you don't have to iterate over all bundle in the runtime.

Reply:
I don't think this is really relevant, this method is actually nowhere used except in this place and one another on startup/shutdown so I'd like to make it as simple and understandable as possible without a chance to break because we interpret what "timestamp" means differently, e.g. the `ModuleDatabase.getRevisionsTimestamp()` says:

>  **For example:**
> 	 * <ul>
> 	 * <li>installing a module
> 	 * <li>updating a module
> 	 * <li>uninstalling a module
> 	 * <li>modifying the wirings
> 	 * </ul>
> 	 * <p>

so this example already contain something we don't want (change in wiring), and "for example" indicates to me that in the future maybe other things will become relevant.

Reply:
> The performance might be secondary, but the suggested approaches also ensure it works as before, because that is what was effectively called by the PlatformAdmin.

Yes an "works like before" is wrong, as the method only want to have a different timestmap if a bundle was installed/updated/uninstalled, nothing else.